packages/roosterjs-editor-plugins/lib/CustomReplace/CustomReplace.ts

```diff
@@ -0,0 +1,157 @@
+    }
+
+    updateReplacements(newReplacements: Replacement[]) {
+        this.replacements = newReplacements;
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    /**
+     * Get a friendly name of  this plugin
+     */
+    getName() {
+        return 'CustomReplace';
+    }
+
+    /**
+     * Initialize this plugin
+     * @param editor The editor instance
+     */
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose(): void {
+    }
+
+    public onPluginEvent(event: PluginEvent) {
+        if (this.editor.isInIME() || event.eventType != PluginEventType.KeyPress) {
+            return;
+        }
+
+        // This handler fires before the event's native behaviour -- defer until after it completes
+        setTimeout(() => this.handlePluginEvent(event), 0)
+    }
+
+    private handlePluginEvent(event: PluginKeyPressEvent) {
+        // Get the matching replacement
+        const range = document.getSelection().getRangeAt(0);
+        if (range == null) return;
+        const searcher = this.editor.getContentSearcherOfCursor();
+        const stringToSearch = searcher.getSubStringBefore(this.longestReplacementLength);
+
+        // Do nothing on control / navigation keys (e.g. keys that did not just type the last character)
+        if (!stringToSearch.endsWith(event.rawEvent.key)) {
+            return
+        }
+
+        const replacement = this.getMatchingReplacement(stringToSearch);
+        if (replacement == null) return;
+
+        // Reconstruct a selection of the text on the document that matches the
+        // replacement we selected.
+        const matchingText = searcher.getSubStringBefore(replacement.sourceString.length)
+        const matchingRange = searcher.getRangeFromText(matchingText, true /* exactMatch */);
+
+        // parse the html string off the dom and inline the resulting element.
+        const parsingSpan = document.createElement('span');
+        parsingSpan.innerHTML = replacement.replacementHTML;
+        const nodeToInsert = (parsingSpan.childNodes.length == 1) ? parsingSpan.childNodes[0] : parsingSpan;
+
+        // Switch the node for the selection range
+        this.editor.addUndoSnapshot(() => {
+            matchingRange.deleteContents();
+            matchingRange.insertNode(nodeToInsert);
+            if (nodeToInsert instanceof Text) {
+                // Fix backspace after text node insertion
+                // (when collapsing the range to the end, the selection goes _after_ the text
+                // node, which lets it delete half of the unicode multipart heart.)
+                const endOfTextNodeRange = new Range();
+                endOfTextNodeRange.setStart(nodeToInsert, nodeToInsert.length);
+                endOfTextNodeRange.setEnd(nodeToInsert, nodeToInsert.length);
+                this.editor.select(endOfTextNodeRange);
```

Comment:
try this.editor.select(nodeToInsert, PositionType.End) rather than creating a range.